### ISSUE TITLE
[Question] Support UI/Menu PSB?

### ISSUE THREAD

**Issue by 0xh3LL0** [UlyssesWu/FreeMote#9]
1. what game are you test on?
2. does it support non-emote? if yes, will you add support on it? cuz in my case, this tools only compatible with [Dracu-Riot](https://vndb.org/v8213), but not with other kirikiri games... even though they have the same format (PSB)

**Comment by UlyssesWu** [comment]
This lib is called Fre**eMote** rather than PsbTools so you can imagine it just focused on emote PSB. I assume M2 uses PSB not only for emote but also for packing scripts (`.ks.psb`) or textures in their games, and these PSBs are **slightly** different.

1. I have tested on 3~4 emote games (from PSBv2 to PSBv4) and all PSBs can be supported. The only thing you should be aware is finding correct `CryptKey` for PSBs and convert them to _pure_ PSBs. All tools except `EmoteConv` assume the input as [_pure_ PSB](https://github.com/Project-AZUSA/FreeMote-SDK).
While I have developed a more advanced tool (`FreeMote.Purify`) to infer the key, it's not going to be released for a while.
For now FreeMote just can not support some texture formats (I will add more later) so it can not produce correct pictures in some cases but you can just export raw bytes and covert them by yourself.
**If you have some PSBs which can not be supported, send them to me by email.**

2. I don't play galgames very much. AFAIK, the PSB which packs scripts (`.ks.psb`) can be supported. The only problem (maybe) is it will be converted to json format rather than those script(ks?) formats. But when you got the json, you can write a tool to convert it as you want.

**Comment by 0xh3LL0** [comment]
oh, i see...
thats why its not compatible with normal PSB because its focused on emote PSB
guess thats the reason, eh... but the decompiler run just fine though...

thanks for answering, and sure i'll email you when i have one!

**Comment by UlyssesWu** [comment]
You can just send me one even if it's not an _emote_ PSB file. Maybe I can tell you what's the problem (if I know).
And if you are working on a krkr game, you should always try [KrkrExtract](https://github.com/xmoeproject/KrkrExtract).

**Comment by 0xh3LL0** [comment]
both psbdecompile and psbuild run just fine on script only, but not the UI / menu...
psbdecompile can extract them to PNG but i cant compile them back with psbuild... i can, however, only on one game i mentioned above, 
i use krkrextract too for pack/unpack the xp3, but it cant pack/compile the UI which is PSB

anyway, i'll email you both the script and UI

**Comment by UlyssesWu** [comment]
Thanks. I have received the file and I'll look into it.

**Comment by 0xh3LL0** [comment]
thank you!
i'll send you more if you need it!

**Comment by UlyssesWu** [comment]
The current commit is able to decompile/compile pimg/scn psb files, but it won't convert `tlg` images (just export/import them). If you can convert `tlg` images by yourself, you can grab the latest build [here](https://ci.appveyor.com/project/UlyssesWu/freemote/build/artifacts). If you wish `FreeMote` convert `tlg` images for you, there is still a lot to be done.

**Comment by 0xh3LL0** [comment]
thank you very much!
i don't have a problem with tlg, as long as the psbuild can compile them back.
i try it just now and the pimg seem can be compiled with psbuild,
i haven't test it yet with the game, but i hope it works,
will tell you the details later.

**Comment by 0xh3LL0** [comment]
yay! it works! thank you very much!
<a href="https://imgur.com/Dew7vBn"><img src="https://i.imgur.com/Dew7vBn.png" title="source: imgur.com" /></a>

**Comment by UlyssesWu** [comment]
Nice 👍 
If it doesn't work for some games, firstly try specific version in PsBuild like `PsBuild /v2`. 
And if there are strange PSBs which are not supported, send it to me via email.

**Comment by 0xh3LL0** [comment]
sure i will, thank you!

**Comment by UlyssesWu** [comment]
BTW, what's the tool you used to convert normal pictures to TLG?

**Comment by 0xh3LL0** [comment]
the tools is [krdevui](https://github.com/krkrz/krdevui), the compiled binary is [here](https://tlwiki.org/images/6/6d/KiriKiri_Official_Tools_English_Translation.zip)

**Comment by 0xh3LL0** [comment]
and since your tools is based on number201724 psbfile, maybe you can use his [text tool](https://github.com/number201724/psbfile/tree/master/texts)?
if that possible, that would be solve your problem on adding ks/scn support...

**Comment by UlyssesWu** [comment]
We can already get and set texts using FreeMote libs, even without necessary to decompile/recompile json (Load PSB -> Modify strings in PSB -> Build PSB. No json required) ([see also](https://github.com/marcussacana/KrKrZSceneManager/issues/1)). That card in Project means a feature to convert scn PSB to a script-like text file (similar to what krkr projects use) rather than json file in order to be used in krkr projects. But I'm not going to implement that actually.

**Comment by 0xh3LL0** [comment]
yes sure, but marcussacana ever said something like [this](https://forums.fuwanovel.net/topic/18136-looking-for-technical-advice-with-the-kirikiri-engine/?do=findComment&comment=460877) so idk, maybe if he can implement those string sort (rather than in alphabetical order), i'll use his tool instead.
anyway i use both his and number201724 tools, haven't encountered any problem yet...

**Comment by UlyssesWu** [comment]
Since `FreeMote.Psb` is a complete PSB parsing lib (`psbfile` can not support PSBv3 encryption and PSBv4, but that doesn't really matter for pimg/scn PSBs), the "string sort" is able to be implemented just in [one line](https://github.com/UlyssesWu/FreeMote/blob/master/FreeMote.Psb/Psb.cs#L448). Using it to implement a tool like `KrKrZSceneManager` is much easier for both developers and users, and should be faster than de/recompile json.
Since I just write and use this lib for emote PSB, I don't have plans to write tools like that for now, but I believe it can be done with FreeMote. And I don't know if FreeMote really supports all kinds of PSB besides emote/pimg/scn PSB.

**Comment by 0xh3LL0** [comment]
yea, hope marcussacana can add those feature on his krkrzcsenemanager soon.

wish the best for you two!

**Comment by 0xh3LL0** [comment]
i just notice that it can convert the TLG to PNG now with /ep, but why it keep the original TLG? doesnt it better if it just automatically deleted, and leave the PNG only?

**Comment by UlyssesWu** [comment]
Because we can't encode TLG. TLG->PNG is ok, but PNG->TLG is not implemented. The decode code is borrowed from a powerful extract tool [GARbro](https://github.com/morkt/GARbro) (MIT License).
So PsBuild still needs `tlg` to compile. Maybe you can convert PNG to TLG with other tools, but some users may just want to change some fields in json and keep original tlgs.

**Comment by 0xh3LL0** [comment]
> some users may just want to change some fields in json and keep original tlgs

man, i didnt even think that far... but hm, that make sense... i see, i see...

**Comment by 0xh3LL0** [comment]
nice...
i'll try it later, and if this works i'll update the tutorial as well...

**Comment by 0xh3LL0** [comment]
confirmed, it works!
